### ISSUE TITLE
"must return value on all code paths" when checking literal field value

### ISSUE THREAD

**Issue by Nhfiam** [microsoft/pyright#7591]
Expected behavior: no pyright errors
Actual behavior: `error: Function with declared return type "int" must return value on all code paths`

**Code**
```python
import typing


class T1:
    x: typing.Literal[1] = 1


class T2:
    x: typing.Literal[2] = 2


def f(t: T1 | T2) -> int:
    if t.x == 1:
        return 0
    if t.x == 2:
        return 0
```

**Version**
`pyright 1.1.356`, command-line tool

**Comment by Nhfiam** [comment]
For this code everything works as expected
```python
import typing


T1 = typing.Literal[1]
T2 = typing.Literal[2]


def f(t: T1 | T2) -> int:
    if t == 1:
        return 0
    if t == 2:
        return 0
```

**Comment by Nhfiam** [comment]
mypy is fine with both examples

**Comment by erictraut** [comment]
This behavior is by design, so I don't consider this a bug. This pattern requires a special type of narrowing called [Narrowing for Implied Else](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=narrowing-for-implied-else). This technique is expensive. When I first implemented the algorithm, it had a significant negative impact on analysis times and latencies for LSP features. I managed to make it perform well for common cases, but I needed to impose limitations on which patterns it supports. One such limitation is that it works only on expressions that are simple names (like `t` in your second example), not attribute access expressions (like `t.x` in your first example). These limitations are spelled out in the linked documentation. 

You'll need to find a workaround if you want this to type check without errors.
